accepted/2022/enable-512-vectors.md

```diff
@@ -0,0 +1,518 @@
+
+The following example shows how `VectorMask128` allows to perform the same functionality without dealing with these lower level details:
+
+```C#
+Vector128<ushort> vector = Vector128.LoadUnsafe(ref source, offset);
+VectorMask128<ushort> v1Eq = Vector128.EqualsByVectorMask(vector, v1);
+VectorMask128<ushort> v2Eq = Vector128.EqualsByVectorMask(vector, v2);
+VectorMask128<ushort> v3Eq = Vector128.EqualsByVectorMask(vector, v3);
+VectorMask128<ushort> cmp = (v1Eq | v2Eq | v3Eq);
+
+if (cmp != VectorMask128<ushort>.Zero)
+{
+    // Skip every other bit
+    do
+    {
+        uint elmPos = cmp.FirstIndexOf(true);
+        sepListBuilder.Append((int)(offset + elmPos));
+        cmp = cmp.SetElementCond(elmPos, false);
+    } while (cmp != 0);
+}
+```
+
+`VectorMask128<ushort>` properly links conditional SIMD processing with `Vector128<ushort>`. The JIT best determines how to lower the `VectorMask` API based on available hardware of the system; however, because `VectorMask` encodes conditional length and element type, it allows the JIT to lower the code optimally.
+
+The following lists the new classes we propose;
+
+
+| Class  | Associated `Vector` API | 
+| ------ | ---------- |
+| `VectorMask64<T>`  | `Vector64<T>`  | 
+| `VectorMask128<T>` | `Vector128<T>` | 
+| `VectorMask256<T>` | `Vector256<T>` | 
+| `VectorMask<T>`    | `Vector<T>`    |  
+
+The following lists the methods for `VectorMask<T>` API that allow for combining `VectorMask<T> into complex conditions, e.g., boolean logic:
+
+| Method  |
+| ------ | 
+| `VectorMask<T> VectorMask<T>.And(VectorMask<T>, VectorMask<T>)`  | 
+| `VectorMask<T> VectorMask<T>.Or(VectorMask<T>, VectorMask<T>)`  | 
+| `VectorMask<T> VectorMask<T>.Not(VectorMask<T>, VectorMask<T>)`  | 
+| `VectorMask<T> VectorMask<T>.Xor(VectorMask<T>, VectorMask<T>)`  | 
+
+The following lists the methods that allow to use `VectorMask<T>` for some lower-level processing typically done in SIMD code
+
+(OPEN: What else?)
+
+| Method  |
+| ------ | 
+| `VectorMask<T> VectorMask<T>.FirstIndexOf(VectorMask<T> mask, bool val)`  | 
+| `VectorMask<T> VectorMask<T>.SetElemntCond(VectorMask<T> mask, ulong pos, bool val)` | 
+
+
+#### Conditional Processing API Methods
+
+As `VectorMask<T>` allows to build conditional logic over SIMD `Vector`, we propose the following additions to the `Vector` API that allow to related one vector to another:
+
+| Method  |
+| ------ | 
+| `VectorMask<T> Vector<T>.EqualsByVectorMask(Vector<T> v1, Vector<T> v2)`  | 
```

Comment:
If we went with the `AsMask()` proposal from above, then these just become `Vector<T>.Equals(x, y).AsMask()`, which while slighlty more verbose doesn't require an "explosion" of new APIs and can be more easily integrated into existing code relying on `ExtractMostSignificantBits` and `MoveMask`

Reply:
From an API standpoint I think this is fine and intuitive, so long as the JIT can potentially optimize `Vector<T>.Equals(x, y).AsMask()` into something that takes advantage of the most performant masking features available, e.g., masking registers for AVX512, e.g., `Vector<T>.Equals(x,y).AsMask()` can be lowered to `vpcmpX` if available.

Reply:
Yep, definitely. We also do and rely on this for a couple similar cases like `ToScalar` today so I don't think it will be particularly problematic.